### PR TITLE
Fix some heap buffer overflows

### PR DESCRIPTION
**Pull request by c0nk**

### CODE REVIEW

**Comment by c0nk** [comment]
@nygard please review

File: Source/CDLCSegment.m [hunk 1]
```diff
@@ -37,7 +37,7 @@ - (id)initWithDataCursor:(CDMachOFileDataCursor *)cursor;
         _segmentCommand.cmdsize = [cursor readInt32];
         
         _name = [cursor readStringOfLength:16 encoding:NSASCIIStringEncoding];
-        memcpy(_segmentCommand.segname, [_name UTF8String], sizeof(_segmentCommand.segname));
+        strncpy(_segmentCommand.segname, [_name UTF8String], sizeof(_segmentCommand.segname) - 1);
```
Comment (Quuxplusone) [inline Source/CDLCSegment.m]: If the original bug was that `memcpy` doesn't null-terminate, then the bug is still there: `strncpy` won't null-terminate either. Strongly consider using `snprintf` instead.

Comment (Quuxplusone) [inline Source/CDLCSegment.m]: Also, do you have to worry about UTF-8 issues? E.g. if the `sizeof(_segmentCommand.segname) - 1`th byte comes in the middle of a UTF-8 byte sequence such that this truncation results in a malformed UTF-8 string?

**Comment by c0nk** [comment]
The original bug was that memcpy was over-running the buffer returned by [_name UTF8String]. There is currently no issue with null-termination. Subtracting -1 from the size of the array ensures the last byte (which is initialized to 0) will stay unchanged.

You are right about broken UTF-8 if the section names can even contain non-ascii characters. In practice does this even happen?

**Comment by c0nk** [comment]
Ping?

**Comment by c0nk** [comment]
Please apply.

**Comment by stek29** [comment]
@nygard Ping

**Comment by nygard** [comment]
Thanks for the pull request.  These should be fixed by another pull request that I recently merged.

**Pull request closed without merging**